executor: Add Oslat test run

RamLavi opened the pull request:
This PR introduces the Oslat test run to the executor library.

RamLavi committed:
executor: Pass Oslat input parameters from configmap

Signed-off-by: Ram Lavi <ralavi@redhat.com>

RamLavi committed:
oslat: New oslat client package with Run method

Introducing a package that will handle the oslat test [0], and the run
method used to run the test.
The Run command runs the oslat test inside the VMI console.
In order to make sure that the test duration does not exceed the checkup
timeout, passing the checkup ctx and wrapping the run command.
The output is checked for successful return value and then parsed for
the
maximum latency value out of the cores that are tested is returned.

the oslat command is run under the taskset command that makes sure that
it runs only on the specified cores under test.
flags used on the oslat command [0]:
- cpu-list: vCPUs under test.
- rtprio: SCHED_FIFO priority (1-99).
- duration: oslat test duration.
- workload: kind of workload tested.
- workload-mem: Size of the memory to use for the workload.

[0] https://www.mankier.com/8/oslat

Signed-off-by: Ram Lavi <ralavi@redhat.com>

RamLavi committed:
executor: Run Oslat test and return maximum latency result

Signed-off-by: Ram Lavi <ralavi@redhat.com>

RamLavi commented:
passed on CNV 4.15 cluster, using my private vm image `quay.io/ramlavi/kubevirt-realtime-checkup:devel`:
```
make e2e-test
mkdir -p /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache
podman run --rm \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup:/go/src/github.com/kiagnose/kubevirt-realtime-checkup:Z \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache:/root/.cache/go-build:Z \
	-v /home/ralavi/.kube/sno03-cnvqe2-rdu2:/root/.kube:Z,ro \
	--workdir /go/src/github.com/kiagnose/kubevirt-realtime-checkup \
	-e KUBECONFIG=/root/.kube/kubeconfig \
	-e TEST_NAMESPACE=realtime-checkup-1 \
	-e TEST_CHECKUP_IMAGE=quay.io/ramlavi/kubevirt-realtime-checkup:devel \
	-e VM_UNDER_TEST_CONTAINER_DISK_IMAGE=quay.io/ramlavi/kubevirt-realtime-checkup-vm:latest \
	docker.io/library/golang:1.20.12 \
	go test -v ./tests/... -test.v -test.timeout=1h -ginkgo.v -ginkgo.timeout=1h 
=== RUN   TestTests
Running Suite: Tests Suite - /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests
========================================================================================
Random Seed: 1702558502

Will run 1 of 1 specs
------------------------------
[BeforeSuite] 
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/tests_suite_test.go:39
[BeforeSuite] PASSED [0.008 seconds]
------------------------------
Checkup execution should complete successfully
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:80
• [745.874 seconds]
------------------------------

Ran 1 of 1 Specs in 745.883 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestTests (745.89s)
PASS
ok  	github.com/kiagnose/kubevirt-realtime-checkup/tests	745.918s
```

logs:
```
$ oc logs -f job/rt-checkup
2023/12/14 12:55:09 kubevirt-realtime-checkup starting...
2023/12/14 12:55:09 Using the following config:
2023/12/14 12:55:09 	"vmUnderTestTargetNodeName": ""
2023/12/14 12:55:09 	"vmUnderTestContainerDiskImage": "quay.io/ramlavi/kubevirt-realtime-checkup-vm:latest"
2023/12/14 12:55:09 	"oslatDuration": "10m0s"
2023/12/14 12:55:09 	"oslatLatencyThresholdMicroSeconds": "45µs"
2023/12/14 12:55:09 Waiting for VMI "realtime-checkup-1/rt-vmi-l4bbn" to boot...
2023/12/14 12:56:14 VMI "realtime-checkup-1/rt-vmi-l4bbn" had successfully booted
2023/12/14 12:56:14 Login to VMI under test...
2023/12/14 12:57:15 Running Oslat test on VMI under test for 10m0s...
2023/12/14 13:07:17 Oslat test completed:
taskset -c 1-2 oslat --cpu-list 1-2 --rtprio 1 --duration 10m0s --workload memmove --workload-mem 4K 
oslat V 2.60
Total runtime: 		600 seconds
Thread priority: 	SCHED_FIFO:1
CPU list: 		1-2
CPU for main thread: 	0
Workload: 		memmove
Workload mem: 		4 (KiB)
Preheat cores: 		2

Pre-heat for 1 seconds...
Test starts...
Test completed.

        Core:	 1 2
Counter Freq:	 2096 2096 (Mhz)
    001 (us):	 0 0
    002 (us):	 6148641856 6192935687
    003 (us):	 0 1
    004 (us):	 1305 1263
    005 (us):	 291 315
    006 (us):	 31 32
    007 (us):	 2 13
    008 (us):	 4 2
    009 (us):	 2 5
    010 (us):	 9 4
    011 (us):	 21 1
    012 (us):	 1 15
    013 (us):	 0 16
    014 (us):	 0 0
    015 (us):	 0 0
    016 (us):	 0 0
    017 (us):	 0 0
    018 (us):	 0 0
    019 (us):	 0 0
    020 (us):	 0 0
    021 (us):	 0 0
    022 (us):	 0 0
    023 (us):	 0 0
    024 (us):	 0 0
    025 (us):	 0 0
    026 (us):	 2 0
    027 (us):	 0 0
    028 (us):	 0 0
    029 (us):	 0 0
    030 (us):	 0 1
    031 (us):	 0 0
    032 (us):	 0 0 (including overflows)
     Minimum:	 1 1 (us)
     Average:	 2.000 2.000 (us)
     Maximum:	 25 29 (us)
     Max-Min:	 24 28 (us)
    Duration:	 599.732 599.732 (sec)

[root@rt-vmi-l4bbn cloud-user]# 
2023/12/14 13:07:17 Max Oslat Latency measured: 29µs
2023/12/14 13:07:17 Trying to delete VMI: "realtime-checkup-1/rt-vmi-l4bbn"
2023/12/14 13:07:17 Waiting for VMI "realtime-checkup-1/rt-vmi-l4bbn" to be deleted...
2023/12/14 13:07:22 VMI "realtime-checkup-1/rt-vmi-l4bbn" was deleted successfully

```

RamLavi commented:
example of failed test (using my private vm image `quay.io/ramlavi/kubevirt-realtime-checkup:devel` and a 5 us threshold): 
```
make e2e-test
mkdir -p /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache
podman run --rm \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup:/go/src/github.com/kiagnose/kubevirt-realtime-checkup:Z \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache:/root/.cache/go-build:Z \
	-v /home/ralavi/.kube/sno03-cnvqe2-rdu2:/root/.kube:Z,ro \
	--workdir /go/src/github.com/kiagnose/kubevirt-realtime-checkup \
	-e KUBECONFIG=/root/.kube/kubeconfig \
	-e TEST_NAMESPACE=realtime-checkup-1 \
	-e TEST_CHECKUP_IMAGE=quay.io/ramlavi/kubevirt-realtime-checkup:devel \
	-e VM_UNDER_TEST_CONTAINER_DISK_IMAGE=quay.io/ramlavi/kubevirt-realtime-checkup-vm:latest \
	docker.io/library/golang:1.20.12 \
	go test -v ./tests/... -test.v -test.timeout=1h -ginkgo.v -ginkgo.timeout=1h 
=== RUN   TestTests
Running Suite: Tests Suite - /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests
========================================================================================
Random Seed: 1702559420

Will run 1 of 1 specs
------------------------------
[BeforeSuite] 
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/tests_suite_test.go:39
[BeforeSuite] PASSED [0.002 seconds]
------------------------------
Checkup execution should complete successfully
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:80
  [FAILED] in [It] - /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:90 @ 12/14/23 13:22:55.745
• [FAILED] [756.202 seconds]
Checkup execution [It] should complete successfully
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:80

  [FAILED] checkup failed: {
  	"spec.param.oslatDuration": "10m",
  	"spec.param.oslatLatencyThresholdMicroSeconds": "5",
  	"spec.param.vmUnderTestContainerDiskImage": "quay.io/ramlavi/kubevirt-realtime-checkup-vm:latest",
  	"spec.param.vmUnderTestTargetNodeName": "",
  	"spec.timeout": "15m",
  	"status.completionTimestamp": "2023-12-14T13:22:49Z",
  	"status.failureReason": "oslat Max Latency measured 30µs exceeded the given threshold 5µs",
  	"status.result.oslatMaxLatencyMicroSeconds": "30",
  	"status.result.vmUnderTestActualNodeName": "",
  	"status.startTimestamp": "2023-12-14T13:10:24Z",
  	"status.succeeded": "false"
  }
  In [It] at: /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:90 @ 12/14/23 13:22:55.745
------------------------------

Summarizing 1 Failure:
  [FAIL] Checkup execution [It] should complete successfully
  /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:90

Ran 1 of 1 Specs in 756.205 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestTests (756.21s)
FAIL
FAIL	github.com/kiagnose/kubevirt-realtime-checkup/tests	756.217s
FAIL
make: *** [Makefile:70: e2e-test] Error 1
```

RamLavi commented:
@orelmisan note that when I'm using the "official" vm image `quay.io/ramlavi/kubevirt-realtime-checkup:devel` then it's seems like oslat is not installed:
```
make e2e-test
mkdir -p /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache
podman run --rm \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup:/go/src/github.com/kiagnose/kubevirt-realtime-checkup:Z \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache:/root/.cache/go-build:Z \
	-v /home/ralavi/.kube/sno03-cnvqe2-rdu2:/root/.kube:Z,ro \
	--workdir /go/src/github.com/kiagnose/kubevirt-realtime-checkup \
	-e KUBECONFIG=/root/.kube/kubeconfig \
	-e TEST_NAMESPACE=realtime-checkup-1 \
	-e TEST_CHECKUP_IMAGE=quay.io/ramlavi/kubevirt-realtime-checkup:devel \
	-e VM_UNDER_TEST_CONTAINER_DISK_IMAGE=quay.io/kiagnose/kubevirt-realtime-checkup-vm:main \
	docker.io/library/golang:1.20.12 \
	go test -v ./tests/... -test.v -test.timeout=1h -ginkgo.v -ginkgo.timeout=1h 
=== RUN   TestTests
Running Suite: Tests Suite - /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests
========================================================================================
Random Seed: 1702560317

Will run 1 of 1 specs
------------------------------
[BeforeSuite] 
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/tests_suite_test.go:39
[BeforeSuite] PASSED [0.002 seconds]
------------------------------
Checkup execution should complete successfully
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:80
  [FAILED] in [It] - /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:90 @ 12/14/23 13:27:02.294
• [FAILED] [106.191 seconds]
Checkup execution [It] should complete successfully
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:80

  [FAILED] checkup failed: {
  	"spec.param.oslatDuration": "10m",
  	"spec.param.oslatLatencyThresholdMicroSeconds": "5",
  	"spec.param.vmUnderTestContainerDiskImage": "quay.io/kiagnose/kubevirt-realtime-checkup-vm:main",
  	"spec.param.vmUnderTestTargetNodeName": "",
  	"spec.timeout": "15m",
  	"status.completionTimestamp": "2023-12-14T13:26:54Z",
  	"status.failureReason": "failed to run Oslat on VMI \"realtime-checkup-1/rt-vmi-g47k4\": oslat test failed with exit code: 127. See logs for more information",
  	"status.startTimestamp": "2023-12-14T13:25:21Z",
  	"status.succeeded": "false"
  }
  In [It] at: /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:90 @ 12/14/23 13:27:02.294
------------------------------

Summarizing 1 Failure:
  [FAIL] Checkup execution [It] should complete successfully
  /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:90

Ran 1 of 1 Specs in 106.194 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestTests (106.19s)
FAIL
FAIL	github.com/kiagnose/kubevirt-realtime-checkup/tests	106.211s
FAIL
make: *** [Makefile:70: e2e-test] Error 1
```
logs:
```
$ oc logs -f job/rt-checkup
2023/12/14 13:25:21 kubevirt-realtime-checkup starting...
2023/12/14 13:25:21 Using the following config:
2023/12/14 13:25:21 	"vmUnderTestTargetNodeName": ""
2023/12/14 13:25:21 	"vmUnderTestContainerDiskImage": "quay.io/kiagnose/kubevirt-realtime-checkup-vm:main"
2023/12/14 13:25:21 	"oslatDuration": "10m0s"
2023/12/14 13:25:21 	"oslatLatencyThresholdMicroSeconds": "5µs"
2023/12/14 13:25:21 Waiting for VMI "realtime-checkup-1/rt-vmi-g47k4" to boot...
2023/12/14 13:26:26 VMI "realtime-checkup-1/rt-vmi-g47k4" had successfully booted
2023/12/14 13:26:26 Login to VMI under test...
2023/12/14 13:26:49 Running Oslat test on VMI under test for 10m0s...
2023/12/14 13:26:49 oslat test returned exit code: 127. stdout: taskset -c 1-2 oslat --cpu-list 1-2 --rtprio 1 --duration 10m0s --workload memmove --workload-mem 4K 
taskset: failed to execute oslat: No such file or directory
[root@rt-vmi-g47k4 cloud-user]# 
2023/12/14 13:26:49 Trying to delete VMI: "realtime-checkup-1/rt-vmi-g47k4"
2023/12/14 13:26:49 Waiting for VMI "realtime-checkup-1/rt-vmi-g47k4" to be deleted...
2023/12/14 13:26:54 VMI "realtime-checkup-1/rt-vmi-g47k4" was deleted successfully
2023/12/14 13:26:54 kubevirt-realtime-checkup failed: failed to run Oslat on VMI "realtime-checkup-1/rt-vmi-g47k4": oslat test failed with exit code: 127. See logs for more information

```

Do you have an idea why the image is not the same as mine? I didn't do anything special AFAIK.

orelmisan commented:
@RamLavi Could you please try the `quay.io/kiagnose/kubevirt-realtime-checkup-vm:main` image again with with the change from PR #47?

RamLavi commented:
> @RamLavi Could you please try the `quay.io/kiagnose/kubevirt-realtime-checkup-vm:main` image again with with the change from PR #47?

@orelmisan it looks like it's working now. what has changed? 
```
$ oc logs -f job/rt-checkup
2023/12/17 08:06:03 kubevirt-realtime-checkup starting...
2023/12/17 08:06:03 Using the following config:
2023/12/17 08:06:03 	"vmUnderTestTargetNodeName": ""
2023/12/17 08:06:03 	"vmUnderTestContainerDiskImage": "quay.io/kiagnose/kubevirt-realtime-checkup-vm:main"
2023/12/17 08:06:03 	"oslatDuration": "10m0s"
2023/12/17 08:06:03 	"oslatLatencyThresholdMicroSeconds": "45µs"
2023/12/17 08:06:03 Waiting for VMI "realtime-checkup-1/rt-vmi-6spqx" to boot...
2023/12/17 08:07:08 VMI "realtime-checkup-1/rt-vmi-6spqx" had successfully booted
2023/12/17 08:07:08 Login to VMI under test...
2023/12/17 08:08:15 Running Oslat test on VMI under test for 10m0s...
2023/12/17 08:18:17 Oslat test completed:
taskset -c 1-2 oslat --cpu-list 1-2 --rtprio 1 --duration 10m0s --workload memmove --workload-mem 4K 
oslat V 2.60
Total runtime: 		600 seconds
Thread priority: 	SCHED_FIFO:1
CPU list: 		1-2
CPU for main thread: 	0
Workload: 		memmove
Workload mem: 		4 (KiB)
Preheat cores: 		2

Pre-heat for 1 seconds...
Test starts...
Test completed.

        Core:	 1 2
Counter Freq:	 2096 2096 (Mhz)
    001 (us):	 0 0
    002 (us):	 6147050441 6193267362
    003 (us):	 1 1
    004 (us):	 105 94
    005 (us):	 43 44
    006 (us):	 7 14
    007 (us):	 2 3
    008 (us):	 1 0
    009 (us):	 3 2
    010 (us):	 15 3
    011 (us):	 12 7
    012 (us):	 1 17
    013 (us):	 0 4
    014 (us):	 0 0
    015 (us):	 0 0
    016 (us):	 0 0
    017 (us):	 0 0
    018 (us):	 0 0
    019 (us):	 0 0
    020 (us):	 0 0
    021 (us):	 0 0
    022 (us):	 0 0
    023 (us):	 0 0
    024 (us):	 0 0
    025 (us):	 0 1
    026 (us):	 0 0
    027 (us):	 0 1
    028 (us):	 0 0
    029 (us):	 0 0
    030 (us):	 0 0
    031 (us):	 0 0
    032 (us):	 0 0 (including overflows)
     Minimum:	 1 1 (us)
     Average:	 2.000 2.000 (us)
     Maximum:	 11 26 (us)
     Max-Min:	 10 25 (us)
    Duration:	 599.731 599.731 (sec)

[root@rt-vmi-6spqx cloud-user]# 
2023/12/17 08:18:17 Max Oslat Latency measured: 26µs
2023/12/17 08:18:17 Trying to delete VMI: "realtime-checkup-1/rt-vmi-6spqx"
2023/12/17 08:18:17 Waiting for VMI "realtime-checkup-1/rt-vmi-6spqx" to be deleted...
2023/12/17 08:18:27 VMI "realtime-checkup-1/rt-vmi-6spqx" was deleted successfully
```

RamLavi commented:
checked after rebased #47 - works!

Thanks @orelmisan. The PR is ready for your review

orelmisan commented on pkg/internal/checkup/executor/oslat/client.go:
```diff
@@ -0,0 +1,195 @@
+			return 0, err
+		}
+	case <-ctx.Done():
+		return 0, errors.NewTimeoutError("oslat test canceled due to checkup timeout", 0)
```
Could you please elaborate what is the reason to use a package from `k8s.io/apimachinery` instead of using the go standard library?

RamLavi replied:
that's just the template I was using, no real need to use it. will change to use the ctx Err
DONE

orelmisan commented on pkg/internal/checkup/executor/oslat/client.go:
```diff
@@ -0,0 +1,195 @@
+	var resp []expect.BatchRes
+	oslatCmd := buildOslatCmd(t.testDuration)
+
+	done := make(chan struct{})
```
Please consider creating a struct to hold `resp` and `err` and pass it using the channel.

RamLavi replied:
sure
DONE

orelmisan commented on pkg/internal/checkup/executor/oslat/client_test.go:
```diff
@@ -0,0 +1,243 @@
+
+	maxLatency, err := oslatTest.Run(context.Background())
+	assert.NoError(t, err, "Run returned an error")
+	expected := 56 * time.Microsecond
```
Its unclear why the actual result is equal to this value.
Please consider explicitly passing it to the expecterStub{}.

RamLavi replied:
DONE

orelmisan commented on pkg/internal/checkup/executor/oslat/client_test.go:
```diff
@@ -0,0 +1,243 @@
+
+		maxLatency, err := oslatTest.Run(context.Background())
+		assert.ErrorContains(t, err, expectedRunErr.Error())
+		assert.Empty(t, maxLatency)
```
Please consider not checking the returned value in case of an error.
Same bellow.

RamLavi replied:
it's convention that the returned value is empty if there is an error.
I prefer to make sure I keep that convention..

orelmisan replied:
If there is an error, the returned value is ignored in most cases.

RamLavi replied:
I just prefer to be explicit, but I don't mind removing if you insist.

orelmisan replied:
I suggest to remove it.
A future reader of the code will be confused IMO about why do we expect a certain value when a function also returns an error.

orelmisan commented on pkg/internal/checkup/executor/oslat/client_test.go:
```diff
@@ -0,0 +1,243 @@
+		"    Duration:\t 59.970 59.970 (sec)\n" +
+		"\n" +
+		"[root@rt-vmi-rw5tr cloud-user]#"
+	oslatRunInvalidOutput = "oslat V 2.60\n" +
```
Please add a new line here in order to separate it from the previous string.

RamLavi replied:
DONE

orelmisan commented on pkg/internal/checkup/checkup.go:
```diff
@@ -87,7 +87,7 @@ func (c *Checkup) Setup(ctx context.Context) error {
 }
 
-func (c *Checkup) Run(ctx context.Context) error {
+func (c *Checkup) Run(ctx context.Context, oslatLatencyThreshold time.Duration) error {
```
The config is passed into `New()`.
Please add a config field to `Checkup` and use it here.
IMO we should not add additional params to `Run()` because it will not scale well.

RamLavi replied:
I figured we can do this later, but let's make our future selves an easier life..
DONE

orelmisan reviewed (changes requested):
Thank you for the PR @RamLavi.
Please see the inline comments.

orelmisan commented on pkg/internal/checkup/executor/executor.go:
```diff
@@ -61,5 +62,15 @@ func (e Executor) Execute(ctx context.Context, vmiUnderTestName string) (status.
+	log.Printf("Max Oslat Latency measured: %s", maxLatency.String())
+
+	return status.Results{
+		OslatMaxLatency: maxLatency,
```
Please an assertion to the E2E test to check that this field is not empty.

RamLavi replied:
DONE

RamLavi commented:
[Change](https://github.com/kiagnose/kubevirt-realtime-checkup/compare/492c580d4b052ef7dfcb1b744a3c781e5a7e22c5..f852674959e9124d2d80b91306540e8b94049154): @orelmisan 's review fixes

RamLavi commented:
Change: [Rebase](https://github.com/kiagnose/kubevirt-realtime-checkup/compare/f852674959e9124d2d80b91306540e8b94049154..07361e2d6cd6a71031aad36408b90fbcb7830e01)

RamLavi committed:
checkup: Add fail logic

The checkup will now fail if the oslat Maximum latency measured exceeded
the oslatLatencyThreshold, and issue the appropriate failure reason.

Signed-off-by: Ram Lavi <ralavi@redhat.com>

RamLavi commented:
[Change](https://github.com/kiagnose/kubevirt-realtime-checkup/compare/07361e2d6cd6a71031aad36408b90fbcb7830e01..21879384a8df67a40b4bc0953ceee667e261c3a3): pass the oslatThreshold via the checkup object, as @orelmisan suggested [here](https://github.com/kiagnose/kubevirt-realtime-checkup/pull/46#discussion_r1429153000)

RamLavi commented:
passed on CNV4.15:
```
make e2e-test
mkdir -p /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache
podman run --rm \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup:/go/src/github.com/kiagnose/kubevirt-realtime-checkup:Z \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache:/root/.cache/go-build:Z \
	-v /home/ralavi/.kube/sno03-cnvqe2-rdu2:/root/.kube:Z,ro \
	--workdir /go/src/github.com/kiagnose/kubevirt-realtime-checkup \
	-e KUBECONFIG=/root/.kube/kubeconfig \
	-e TEST_NAMESPACE=realtime-checkup-1 \
	-e TEST_CHECKUP_IMAGE=quay.io/ramlavi/kubevirt-realtime-checkup:devel \
	-e VM_UNDER_TEST_CONTAINER_DISK_IMAGE=quay.io/kiagnose/kubevirt-realtime-checkup-vm:main \
	docker.io/library/golang:1.20.12 \
	go test -v ./tests/... -test.v -test.timeout=1h -ginkgo.v -ginkgo.timeout=1h 
=== RUN   TestTests
Running Suite: Tests Suite - /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests
========================================================================================
Random Seed: 1702890031

Will run 1 of 1 specs
------------------------------
[BeforeSuite] 
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/tests_suite_test.go:39
[BeforeSuite] PASSED [0.002 seconds]
------------------------------
Checkup execution should complete successfully
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:80
• [821.167 seconds]
------------------------------

Ran 1 of 1 Specs in 821.169 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestTests (821.17s)
PASS
ok  	github.com/kiagnose/kubevirt-realtime-checkup/tests	821.178s
```

logs:
```
$ oc logs -f job/realtime-checkup
2023/12/18 09:00:39 kubevirt-realtime-checkup starting...
2023/12/18 09:00:39 Using the following config:
2023/12/18 09:00:39 	"vmUnderTestTargetNodeName": ""
2023/12/18 09:00:39 	"vmUnderTestContainerDiskImage": "quay.io/kiagnose/kubevirt-realtime-checkup-vm:main"
2023/12/18 09:00:39 	"oslatDuration": "10m0s"
2023/12/18 09:00:39 	"oslatLatencyThresholdMicroSeconds": "45µs"
2023/12/18 09:00:39 Waiting for VMI "realtime-checkup-1/rt-vmi-tvpnb" to boot...
2023/12/18 09:02:54 VMI "realtime-checkup-1/rt-vmi-tvpnb" had successfully booted
2023/12/18 09:02:54 Login to VMI under test...
2023/12/18 09:04:00 Running Oslat test on VMI under test for 10m0s...
2023/12/18 09:14:02 Oslat test completed:
taskset -c 1-2 oslat --cpu-list 1-2 --rtprio 1 --duration 10m0s --workload memmove --workload-mem 4K 
oslat V 2.60
Total runtime: 		600 seconds
Thread priority: 	SCHED_FIFO:1
CPU list: 		1-2
CPU for main thread: 	0
Workload: 		memmove
Workload mem: 		4 (KiB)
Preheat cores: 		2

Pre-heat for 1 seconds...
Test starts...
Test completed.

        Core:	 1 2
Counter Freq:	 2096 2096 (Mhz)
    001 (us):	 0 0
    002 (us):	 6149216625 6192544634
    003 (us):	 1 1
    004 (us):	 108 99
    005 (us):	 40 38
    006 (us):	 8 12
    007 (us):	 1 4
    008 (us):	 1 1
    009 (us):	 2 4
    010 (us):	 11 0
    011 (us):	 14 0
    012 (us):	 1 10
    013 (us):	 1 17
    014 (us):	 0 2
    015 (us):	 0 0
    016 (us):	 0 0
    017 (us):	 0 0
    018 (us):	 0 0
    019 (us):	 0 0
    020 (us):	 0 0
    021 (us):	 0 0
    022 (us):	 0 0
    023 (us):	 0 0
    024 (us):	 0 0
    025 (us):	 0 0
    026 (us):	 0 0
    027 (us):	 0 0
    028 (us):	 0 0
    029 (us):	 0 0
    030 (us):	 0 1
    031 (us):	 0 0
    032 (us):	 0 0 (including overflows)
     Minimum:	 1 1 (us)
     Average:	 2.000 2.000 (us)
     Maximum:	 12 29 (us)
     Max-Min:	 11 28 (us)
    Duration:	 599.704 599.704 (sec)

[root@rt-vmi-tvpnb cloud-user]# 
2023/12/18 09:14:02 Max Oslat Latency measured: 29µs
2023/12/18 09:14:02 Trying to delete VMI: "realtime-checkup-1/rt-vmi-tvpnb"
2023/12/18 09:14:02 Waiting for VMI "realtime-checkup-1/rt-vmi-tvpnb" to be deleted...
2023/12/18 09:14:07 VMI "realtime-checkup-1/rt-vmi-tvpnb" was deleted successfully
```

orelmisan commented on pkg/internal/checkup/executor/oslat/client.go:
```diff
@@ -0,0 +1,200 @@
+	testDuration    time.Duration
+}
+
+func NewTest(vmiUnderTestConsoleExpecter consoleExpecter, testDuration time.Duration) *Client {
```
The function returns a `Client` object.
```suggestion
func NewClient(vmiUnderTestConsoleExpecter consoleExpecter, testDuration time.Duration) *Client {
```

RamLavi replied:
DONE

orelmisan commented on pkg/internal/checkup/executor/oslat/client_test.go:
```diff
@@ -0,0 +1,248 @@
+
+func TestRunSuccess(t *testing.T) {
+	expecter := &expecterStub{
+		injectedActualMaxResults: "27 56 (us)",
```
Please consider taking this value as a `time.Duration`.

RamLavi replied:
I think we should keep to the original string and not do another conversion.  
It's just how the Oslat outputs the results, so it's better to just show it how it is.

orelmisan replied:
Its meaning is not very clear IMO when reading the test.

orelmisan commented on pkg/internal/checkup/executor/executor.go:
```diff
@@ -61,5 +62,15 @@ func (e Executor) Execute(ctx context.Context, vmiUnderTestName string) (status.
 	}
 
-	return status.Results{}, nil
+	oslatTest := oslat.NewTest(vmiUnderTestConsoleExpecter, e.OslatDuration)
```
`oslat.NewTest()` returns `oslat.Client`.
Please consider renaming `oslatTest`.

RamLavi replied:
DONE

orelmisan commented on tests/checkup_test.go:
```diff
@@ -106,6 +106,9 @@ var _ = Describe("Checkup execution", func() {
 			ToNot(BeEmpty(), fmt.Sprintf("vmUnderTestActualNodeName should not be empty %+v", configMap.Data))
+		Expect(configMap.Data["status.result.oslatMaxLatencyMicroSeconds"]).
+			ToNot(BeEmpty(), fmt.Sprintf("oslatMaxLatencyMicroSeconds should not be empty %+v", configMap.Data))
+
```
nit: Redundant newline.

RamLavi replied:
DONE

orelmisan commented on pkg/internal/checkup/checkup.go:
```diff
@@ -102,6 +102,10 @@ func (c *Checkup) Run(ctx context.Context) error {
 	}
 	c.results.VMUnderTestActualNodeName = c.vmi.Status.NodeName
 
+	if c.results.OslatMaxLatency > c.cfg.OslatLatencyThreshold {
```
Please consider squashing this commit with the previous one.

RamLavi replied:
DONE

RamLavi commented:
[change](https://github.com/kiagnose/kubevirt-realtime-checkup/compare/21879384a8df67a40b4bc0953ceee667e261c3a3..d7a318b3dde08d127f497c3fb830bd05836f0423): Second round review

orelmisan commented:
@RamLavi does the E2E test passes?

RamLavi commented:
> @RamLavi does the E2E test passes?

yes
```
make e2e-test
mkdir -p /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache
podman run --rm \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup:/go/src/github.com/kiagnose/kubevirt-realtime-checkup:Z \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache:/root/.cache/go-build:Z \
	-v /home/ralavi/.kube/sno03-cnvqe2-rdu2:/root/.kube:Z,ro \
	--workdir /go/src/github.com/kiagnose/kubevirt-realtime-checkup \
	-e KUBECONFIG=/root/.kube/kubeconfig \
	-e TEST_NAMESPACE=realtime-checkup-1 \
	-e TEST_CHECKUP_IMAGE=quay.io/ramlavi/kubevirt-realtime-checkup:devel \
	-e VM_UNDER_TEST_CONTAINER_DISK_IMAGE=quay.io/kiagnose/kubevirt-realtime-checkup-vm:main \
	docker.io/library/golang:1.20.12 \
	go test -v ./tests/... -test.v -test.timeout=1h -ginkgo.v -ginkgo.timeout=1h 
=== RUN   TestTests
Running Suite: Tests Suite - /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests
========================================================================================
Random Seed: 1702976949

Will run 1 of 1 specs
------------------------------
[BeforeSuite] 
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/tests_suite_test.go:39
[BeforeSuite] PASSED [0.002 seconds]
------------------------------
Checkup execution should complete successfully
/go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests/checkup_test.go:80
• [750.633 seconds]
------------------------------

Ran 1 of 1 Specs in 750.636 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestTests (750.64s)
PASS
ok  	github.com/kiagnose/kubevirt-realtime-checkup/tests	750.646s
```

orelmisan reviewed (approved):
Thank you
